design/20230302.gomod.md

```diff
@@ -0,0 +1,334 @@
+# Design: More Modules
+
+NB: This design doc follows from a Hackathon by @SgtCoDFish and @inteon. The intention here is to describe what we did
+and what we discovered, with an eye to seeking consensus and merging upstream.
+
+## Problem Statement
+
+**In short:** Some of our dependencies are complex which makes them hard to upgrade in already-released versions
+
+Upgrading the dependencies of even simple Go projects can be tricky and for a more complex project like cert-manager
+it can be impossible to upgrade dependencies for older releases while satisfying all constraints that we place on
+ourselves as maintainers.
+
+In our case, these constraints are to:
+
+1. Minimise / eliminate CVE reports for any supported release of cert-manager
+2. Be conservative about upgrades, and avoid major version bumps in already-released software
+
+Since we have one `go.mod` file for all of our built binaries, it's not possible for us to be selective about upgrades,
+either.  If, say, only the `controller` component were to report as having a critical vulnerability, we'd have no
+way of fixing only that one vulnerability while leaving everything else untouched.
+
+Essentially, our current project layout forces us to made difficult choices whenever we need to upgrade things.
+
+### Problem Example
+
+**In short:** An example of how upgrades can be particularly difficult in some cases, with no good options.
+
+At the time of writing, cert-manager 1.10 is still in support and depends on Helm because it's imported by `cmctl` (and
+only `cmctl`). We can see the dependency in [go.mod](https://github.com/cert-manager/cert-manager/blob/f54dd1dc98900607e1db7bd4ac2512f0bfe39301/go.mod#L41).
+
+There's a vulnerability reported for Helm v3.10.3 ([1]) which we'd like to patch, but the only version with a fix
+available is Helm v3.11.1.
+
+Between Helm 3.10 and 3.11, several of Helm's dependencies were upgraded, and crucially Helm has some of the same
+dependencies that cert-manager does. That means that we can't easily _just_ upgrade Helm.
+
+Running `go get -u helm.sh/helm/v3` produces 56 different upgrades of _other_ dependencies. Notably, it bumps our
+Kubernetes dependencies from v0.25.2 to v0.26.0 but there are several other changes.
+
+(NB: Helm is just an example here and we could have problems with any package)
+
+## Proposed Solution: Go Module Proliferation
+
+**In short:** Add several new `go.mod` files so individual components can be patched independently
+
+We can create several new Go modules so that each binary we build can have distinct dependencies. This would mean that
+`cmctl` having a dependency on Helm would only affect `cmctl` and wouldn't force us to change any of the other
+components we build in order to patch a Helm vulnerability.
```

Comment:
I personally think this makes the dependency graph a lot harder to reason about, although I can understand the rationale for it.

I'd also call out the one other place where I've seen this done a lot, is in otel.. importing otel is absolutely horrendous, especially when you import another project that also itself depends on otel (there have been a number of dependency graphs like this that are totally unresolvable and have led to either impossible to resolve situations, or some absolutely horrendous hacks).

Personally, I would advocate for:

1) being wary about dependencies like Helm, which are 'binary dependencies' in a sense (albeit they contain many libraries themselves that are useful, they are not engineered towards being easily importable as a library). Either avoiding depending on them as libraries altogether (i.e. shelling out to their built binary equivalents), or if this is just for cmctl, considering splitting _just_ cmctl out...

2) being a better 'library' ourselves (e.g. having a distinct module for our API types, having generated clientsets be `internal/`)

I think this proposal overall will help somewhat, but it shifts the problem down the line more than anything (what happens when one of our binaries still has a lot of dependencies on a lot of different projects?).

Also, we may find ourselves in a position where we are able to bump say, the k8s dependencies for all components _other than_ cmctl, as cmctl still depends on Helm at a version that prevents us upgrading k8s dependencies. This means we end out with a single cert-manager release potentially depending on multiple different versions of different packages/libraries, depending on what component you are considering.

Just my 2c on it - I like the idea and the motive. I'm a little wary of anything that involves making our dependency graph multi-dimensional :)

Reply:
> being wary about dependencies like Helm...

I think we all are!

> I'd also call out the one other place where I've seen this done a lot, is in otel

otel seems very [different](https://github.com/open-telemetry/opentelemetry-go/blob/795ad971191b0bd7ff0d376a19b00e2e20f3b4b7/go.mod) to me. They have a replace statement for their own package in their main go.mod file, and in their [own imported package's go.mod](https://github.com/open-telemetry/opentelemetry-go/blob/795ad971191b0bd7ff0d376a19b00e2e20f3b4b7/metric/go.mod).

That seems like a hellscape. Nobody is advocating for that in cert-manager. It would make importing cert-manager itself confusing, would make it harder to change cert-manager and it would mean that the code we test would be different from the code people would use when importing.

The only things in this proposal which add replace statements are modules that we don't expect others to import (binaries and integration / e2e tests), and those only have replace statements in one direction.

The main cert-manager package (which we _do_ expect people to import) will only improve as a result of this design - by removing some dependencies from it.

> Also, we may find ourselves in a position where we are able to bump say, the k8s dependencies for all components other than cmctl...

This is sorta discussed above, in reverse - above, we talk about Helm needing to be patched which forces just cmctl to bump its k8s dependencies, leaving everything else "behind".

We may indeed find we're in the situation you describe but I'd say that it's probably better to be able to bump all other components and just leave one behind than to have to leave everything behind because we're blocked by one component.

Of course, nobody would want the versions of our deps to drift within a single release; I can't see a good justification for wanting to do that. My concern is more that we might be put in a situation where we're _forced_ to do that, and I think in that case it's good to at least have that option, which this design allows.